File: leetcode-cn/39.组合总和.py
#
# @lc app=leetcode.cn id=39 lang=python3
#
# [39] 组合总和
#

# @lc code=start
from typing import List


class Solution:
    def combinationSum(self, candidates: List[int],
                       target: int) -> List[List[int]]:
        ans = []
        path = []
        candidates.sort()

        def help(candidates, t):
            if t == 0:
                ans.append(path[:])
                return

            for i, num in enumerate(candidates):
                if num > t:
                    break
                path.append(num)
                help(candidates[i:], t - num)
                path.pop()

        help(candidates, target)

        return ans


# @lc code=end

File: leetcode-cn/C++/39.组合总和.cpp
/*
 * @lc app=leetcode.cn id=39 lang=cpp
 *
 * [39] 组合总和
 */

// @lc code=start
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> ans;

    void help(vector<int> candidates, int index, vector<int>& path, int target)
    {
        if (target == 0) {
            ans.emplace_back(path);
            return;
        }

        for (int i = index; i < candidates.size(); i++) {
            int num = candidates[i];
            if (num > target) {
                break;
            }
            path.push_back(num);
            help(candidates, i, path, target - num);
            path.pop_back();
        }
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target)
    {
        sort(candidates.begin(), candidates.end());

        vector<int> path;
        help(candidates, 0, path, target);

        return ans;
    }
};
// @lc code=end
